Sync cert generation before starting webhooks

yiqigao217 opened the pull request:
Make controller manager start only with the cert generation controllers
if cert rotation is not disabled (by "disable-cert-rotation" flag) and
the webhook is enabled. All other controllers will be added to the
manager on the fly if cert generation is disabled or the cert files are
ready. The webhook will be added then too.

Make 'CertRotator' more robust from cert rotation contention among
multiple pods.

Tested on a GKE cluster. Passed "make test-e2e".

Fixes #465

yiqigao217 commented:
/assign @maxsmythe 
/cc @ritazh

yiqigao217 commented:
One thing I noticed in the log is that the manager still restarts once, but the error is in the `ReconcileVWH` controller. The error msg is "Cert secret is not well-formed, missing ca.crt" and "secret is not well-formed, cannot update ValidatingWebhookConfiguration".

maxsmythe commented on pkg/webhook/certs.go:
```diff
@@ -333,16 +342,16 @@ func lookaheadTime() time.Time {
-func (cr *certRotator) validCACert(cert, key []byte) bool {
-	valid, err := validCert(cert, cert, key, cr.caName, lookaheadTime())
+func (cr *CertRotator) validCACert(cert, key []byte) bool {
+	valid, err := validCert(cert, cert, key, cr.CaName, lookaheadTime())
```
Apparently the go style guide says this should be CAName, sorry I used the wrong casing :/

maxsmythe replied:
or caName

yiqigao217 replied:
Thanks for catching this!

maxsmythe commented on pkg/webhook/certs.go:
```diff
@@ -44,9 +44,10 @@ var crLog = logf.Log.WithName("cert-rotation")
-func AddRotator(mgr manager.Manager, cr *certRotator, vwhName string) error {
+// AddRotator adds the CertRotator and ReconcileVWH to the manager.
+func AddRotator(mgr manager.Manager, cr *CertRotator, vwhName string) error {
 	// Use a new client so we are unaffected by the cache sync kill signal
```
We no longer need to create a second client if we are ensuring that rotation happens before other controllers are added, as we no longer risk a premature tear-down. We should remove this code to save on memory.

yiqigao217 replied:
SG, will remove it.

Is it that this new client won't be affected when the manager stops? So that this client can still write secret in the previous 5-s sleep?

maxsmythe replied:
The manager usually shuts down the client before it exits, which means the 5 second sleep would have never helped. By creating a second client, we bypassed the manager's shut down logic at the cost of duplicating the cache a bit. Now that the manager is no longer tearing down, this should no longer be necessary.

maxsmythe commented on pkg/webhook/certs.go:
```diff
@@ -496,3 +505,30 @@ func (r *ReconcileVWH) Reconcile(request reconcile.Request) (reconcile.Result, e
+		Jitter:   1,
+		Steps:    10,
+	}, checkFn); err != nil {
+		crLog.Error(err, "max retries for checking certs validity")
```
This should be a fatal error that causes the `Start()` function to exit with an error so that the master does not hang indefinitely.

yiqigao217 replied:
Good catch, yes. I added a channel `certsNotMounted` from this goroutine to `Start()` and closes the channel here when max retries is reached.

yiqigao217 commented:
> Regarding the continual restarts. I suspect those errors your seeing are not causing the extra restart, as they are likely from the reconciler's call: `artifacts, err := buildArtifactsFromSecret(secret)`, and reconciliation errors are recoverable, as they simply trigger a retry. Is the webook still complaining about an invalid secret?
> 
The restart only happens once and it's before adding other controllers. Ah, these are the errors before the cert rotation stops and causing the manager to stop:  
```
error: "Operation cannot be fulfilled on secrets "gatekeeper-webhook-server-cert": the object has been modified; please apply your changes to the latest version and try again"   
  level: "error"   
  logger: "cert-rotation"   
  msg: "could not refresh cert on startup" 
```
```
error: "could not refresh cert on startup: Operation cannot be fulfilled on secrets "gatekeeper-webhook-server-cert": the object has been modified; please apply your changes to the latest version and try again"   
  errorVerbose: "Operation cannot be fulfilled on secrets "gatekeeper-webhook-server-cert": the object has been modified; please apply your changes to the latest version and try again
could not refresh cert on startup
github.com/open-policy-agent/gatekeeper/pkg/webhook.(*CertRotator).Start
	/go/src/github.com/open-policy-agent/gatekeeper/pkg/webhook/certs.go:91
sigs.k8s.io/controller-runtime/pkg/manager.(*controllerManager).startLeaderElectionRunnables.func1
	/go/src/github.com/open-policy-agent/gatekeeper/vendor/sigs.k8s.io/controller-runtime/pkg/manager/internal.go:477
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357"   
  level: "error"   
  logger: "setup"   
  msg: "problem running manager" 
```

yiqigao217 commented:
> I'm guessing changing the value of the secret doesn't change the value as reflected in the secret mount instantaneously. Instead of reading from the API server to verify we have a valid cert, it may be worth reading `tls.key` and `tls.cert` from the certDir provided to `main.go`, as this is what the server reads from per:
> 
> https://github.com/kubernetes-sigs/controller-runtime/blob/cfbbd753951b51ca4c02b57a348d2c5b39e69f88/pkg/manager/manager.go#L171-L175

SG, I will make sure the files exist in the mountPath instead. 

BTW it's weird that in the testing of GK and HNC (using GK cert - [initial commit](https://github.com/yiqigao217/multi-tenancy/commit/fdf5919a2f8a123c504bc961c96d71a92355a947)), the manager both restarts once. GK restarts before adding other controllers/webhooks. HNC restarts because the cert files are not found in the mountPath.

yiqigao217 commented:
As mentioned above, I tested the initial PR before addressing the review comments both in GK and HNC. Both restarted once:
* GK got errors `Operation cannot be fulfilled on secrets "gatekeeper-webhook-server-cert": the object has been modified; please apply your changes to the latest version and try again` before adding other controllers/webhooks.
* HNC got errors when starting webhook `open /tmp/k8s-webhook-server/serving-certs/tls.crt: no such file or directory`

After addressing the comments (specifically switching from checking the existence of `secret` to cert files in `mountPath`), the test in HNC got 0 restarts as expected and I can see the logs that keeps checking the existence of cert files before starting webhooks. 

However, the test in GK still had the same 1 restart and the same error `Operation cannot be fulfilled on secrets "gatekeeper-webhook-server-cert"..`

yiqigao217 commented on pkg/webhook/certs.go:
```diff
@@ -151,15 +162,15 @@ func (cr *certRotator) refreshCerts(refreshCA bool, secret *corev1.Secret) (bool
 	}
 	writeFn := func() (bool, error) {
 		if err := cr.writeSecret(cert, key, caArtifacts, secret); err != nil {
-			return false, err
```
I think it was a bug to return the `err` here, since it will abort the backoff loop. I changed it to `nil` if the secret was not able to be updated, so it will backoff and retry.

maxsmythe replied:
Good call.

I think the retry with exponential backoff needs to be around refreshCertsIfNeeded, as the value of `refreshCA` could change depending on the value of the re-read secret.

yiqigao217 replied:
Ah right! I added the backoff to `refreshCertsIfNeeded` PTAL.

yiqigao217 commented:
I made a minor change to `writeFn` to write secrets with backoff. The manager now has 0 restarts! However, it seems that I got 1 out of 12 e2e tests failed:
```
not ok 12 required labels audit test
# (from function `wait_for_process' in file test/bats/helpers.bash, line 65,
#  in test file test/bats/test.bats, line 133)
#   `wait_for_process $WAIT_TIME $SLEEP_TIME "kubectl get k8srequiredlabels.constraints.gatekeeper.sh ns-must-have-gk -o json | jq '.status.violations[]'"' failed
# jq: error (at <stdin>:43): Cannot iterate over null (null)
# jq: error (at <stdin>:43): Cannot iterate over null (null)
...(repeating the above error)...
```

maxsmythe commented on main.go:
```diff
@@ -134,6 +147,86 @@ func main() {
 		os.Exit(1)
```
I'm wondering if the reason the audit test is failing is that there is a deadlock here when the webhook is not enabled? That or a fight between the two admission controllers that causes one to never report success? We should figure out that root cause.

After that, we should may sure we are gating cert generation both on whether we have disabled cert rotation and whether the webhook is enabled (`operations["webook"]` evaluates to `true`)

maxsmythe replied:
I think the reason the audit pod never starts is because the secret is never mounted, so the cert bootstrap check never passes.

yiqigao217 replied:
The cert generation only depends on `disableCertRotation` flag. I noticed, similar to the 'manager' pod, the 'audit' pod may or may not have restarts. The issue is in the `writeFn()` in updating the secret.

Also, the audit test fails even when the audit pod is already running.

maxsmythe replied:
I'm guessing this comment is no longer relevant?

yiqigao217 replied:
Right, since we enable certRotation only when both `disableCertRotation` and `operation["webhook"]` meet the condition.

yiqigao217 commented on pkg/webhook/certs.go:
```diff
@@ -151,15 +162,15 @@ func (cr *certRotator) refreshCerts(refreshCA bool, secret *corev1.Secret) (bool
 	writeFn := func() (bool, error) {
 		if err := cr.writeSecret(cert, key, caArtifacts, secret); err != nil {
-			return false, err
+			return false, nil
 		}
 		return true, nil
 	}
 	if err := wait.ExponentialBackoff(wait.Backoff{
-		Duration: 10 * time.Millisecond,
+		Duration: 30 * time.Millisecond,
 		Factor:   2,
 		Jitter:   1,
-		Steps:    10,
+		Steps:    5,
 	}, writeFn); err != nil {
 		log.Error(err, "max retries for writing of secret reached")
 		return false, err
```
The reason for restarts is that the secret is not able to update here. It may restart once and succeed to write on the restart (in previous tests for the "manager" pod case). It may not be able to write and keep restarting (the "audit" pod case). The errors are the same: `“Operation cannot be fulfilled on secrets "gatekeeper-webhook-server-cert": the object has been modified; please apply your changes to the latest version and try again`

yiqigao217 commented:
As discussed offline with @maxsmythe , this latest push fixed the restart issue and the previously failed e2e test by:
* Add additional `operation["webhook"]` besides `!*disableCertRotation` to the condition to add `CertRotator` to the manager. Since when cert rotation is enabled, we now check the existence of the cert files in the `mountPath` instead of the existence of the secret and the `audit` pod doesn't have `volumeMounts`, this will cause `audit` pod to keep restarting.
* Add secret read and validation before refreshing (write) secret in `certRotator`. This makes `certRotator` more robust to contention from multiple pods using the same `certRotator`.

Therefore, now, both the manager and the audit pods have 0 restarts and passed all e2e tests.

yiqigao217 commented:
The build is failed here but it worked on my local machine with `make docker-build REGISTRY=gcr.io/yiqigao-gke-dev`.

Here's the failure in the build:
```
FAIL	github.com/open-policy-agent/gatekeeper/pkg/controller/constrainttemplate	4.844s
ok  	github.com/open-policy-agent/gatekeeper/pkg/controller/sync	0.011s	coverage: 12.3% of statements
?   	github.com/open-policy-agent/gatekeeper/pkg/logging	[no test files]
ok  	github.com/open-policy-agent/gatekeeper/pkg/metrics	0.113s	coverage: 25.0% of statements
ok  	github.com/open-policy-agent/gatekeeper/pkg/target	1.523s	coverage: 67.9% of statements
?   	github.com/open-policy-agent/gatekeeper/pkg/upgrade	[no test files]
ok  	github.com/open-policy-agent/gatekeeper/pkg/util	0.011s	coverage: 43.4% of statements
ok  	github.com/open-policy-agent/gatekeeper/pkg/util/constraint	0.013s	coverage: 72.2% of statements
ok  	github.com/open-policy-agent/gatekeeper/pkg/watch	7.859s	coverage: 66.2% of statements
ok  	github.com/open-policy-agent/gatekeeper/pkg/webhook	4.018s	coverage: 40.0% of statements
FAIL
make: *** [Makefile:65: native-test] Error 1
make: *** [test] Error 2
Makefile:70: recipe for target 'test' failed
##[error]Process completed with exit code 2.
```

maxsmythe commented on pkg/webhook/certs.go:
```diff
@@ -248,15 +262,15 @@ func buildArtifactsFromSecret(secret *corev1.Secret) (*KeyPairArtifacts, error)
-			CommonName:   caName,
-			Organization: []string{cr.caOrganization},
+			CommonName:   cr.CAName,
+			Organization: []string{cr.CaOrganization},
```
nit: CAOrganization

yiqigao217 replied:
Thanks updated!

sozercan commented on pkg/webhook/certs.go:
```diff
@@ -496,3 +510,27 @@ func (r *ReconcileVWH) Reconcile(request reconcile.Request) (reconcile.Result, e
 }
+
+// ensureCertsExist ensure the cert files exist.
+func (cr *CertRotator) ensureCertsExist() {
```
can we add a test for this?

yiqigao217 replied:
Thanks for the review! 

This function is an IPC from cert generation to adding controllers (specifically webhooks) to the controller manager on the fly after it's started. It checks the file existence in the mountPath. 

I am not sure how to test `os.Stat(certFile)`. What I did is a manual test in a deployment. Please let me know if it's good enough or do you have any suggestions on the testing?

sozercan replied:
is it possible to do something like this https://talks.golang.org/2012/10things.slide#8 to mock the filesystem?

maxsmythe replied:
Is there a specific case you're interested in unit testing? Right now we have two logic paths:

File exists -> close happy channel
File missing -> close sad channel

sozercan replied:
sure, if its not worth the effort we can skip the unit test

yiqigao217 replied:
Hi @sozercan , do you mean it's okay to skip this unit test? If so, could you please review the change and merge it if it looks fine for you? Thanks!

sozercan replied:
@yiqigao217 approved, can you rebase and then we can merge. thanks!

yiqigao217 replied:
Sure thing! Done and the pretests are passed! Thanks!
